src/codegen/mod.rs

```diff
@@ -0,0 +1,2000 @@
+                    unnamed_arguments += 1;
+                    format!("arg{}", unnamed_arguments)
+                }
+            };
+
+            assert!(!arg_name.is_empty());
+
+            ast::Arg {
+                ty: arg_ty,
+                pat: aster::AstBuilder::new().pat().id(arg_name),
+                id: ast::DUMMY_NODE_ID,
+            }
+        }).collect::<Vec<_>>();
+
+        let decl = P(ast::FnDecl {
+            inputs: arguments,
+            output: ret,
+            variadic: self.is_variadic(),
+        });
+
+        let fnty = ast::TyKind::BareFn(P(ast::BareFnTy {
+            unsafety: ast::Unsafety::Unsafe,
+            abi: self.abi(),
+            lifetimes: vec![],
+            decl: decl,
+        }));
+
+        P(ast::Ty {
+            id: ast::DUMMY_NODE_ID,
+            node: fnty,
+            span: ctx.span(),
+        })
+    }
+}
+
+impl CodeGenerator for Function {
+    type Extra = Item;
+
+    fn codegen(&self,
+               ctx: &BindgenContext,
+               result: &mut CodegenResult,
+               item: &Item) {
+        let name = self.name();
+        let canonical_name = item.canonical_name(ctx);
+
+        // TODO: Maybe warn here if there's a type/argument mismatch, or
+        // something?
+        if result.seen_function(&canonical_name) {
+            return;
+        }
+        result.saw_function(&canonical_name);
+
+        let signature_item = ctx.resolve_item(self.signature());
+        let signature = signature_item.kind().expect_type();
+        let signature = match *signature.kind() {
+            TypeKind::Function(ref sig) => sig,
+            _ => panic!("How?"),
+        };
+
+        let fndecl = utils::rust_fndecl_from_signature(ctx, signature_item);
+
+        let mut attributes = vec![];
+
+        if let Some(comment) = item.comment() {
+            attributes.push(doc!(comment));
+        }
+
+        if let Some(mangled) = self.mangled_name() {
+            attributes.push(link_name!(mangled));
+        } else if name != canonical_name {
+            attributes.push(link_name!(name));
+        }
+
+        let foreign_item_kind =
+            ast::ForeignItemKind::Fn(fndecl, ast::Generics::default());
+
+        let foreign_item =
+            ast::ForeignItem {
+                ident: ctx.rust_ident_raw(&canonical_name),
+                attrs: attributes,
+                node: foreign_item_kind,
+                id: ast::DUMMY_NODE_ID,
+                span: ctx.span(),
+                vis: ast::Visibility::Public,
+            };
+
+        let item = ForeignModBuilder::new(signature.abi())
+            .with_foreign_item(foreign_item)
+            .build(ctx);
+
+        result.push(item);
+    }
+}
+
+type ItemSet = BTreeSet<ItemId>;
+
+trait TypeCollector {
+    type Extra;
+
+    fn collect_types(&self,
+                     context: &BindgenContext,
+                     types: &mut ItemSet,
+                     extra: &Self::Extra);
```

Comment:
At first I thought so, but that implies either parameterizing the trait and type on a lifetime and whatnot, or pass `Item`s by value, which is both inefficient and not the intended thing. I tried to keep this simple since `extra` uses to be just the `Item` that refers to us, but if that changes we can refactor the interface.